text/0031-integ-testing.md

```diff
@@ -0,0 +1,300 @@
+
+repoTest.assertAws.
+  sqs.receiveMessage({ queue: stack.publishNotifs })
+     .returns({ Messages: [ ... ] });
+```
+
+### Assertions
+
+`contest` comes canned with a set of commonly used high level assertions.
+
+* `assertAws`: execute AWS service APIs and assert the response.
+* `invokeLambda`: invoke an AWS lambda function and assert the response.
+* `curlApi`: Execute the 'curl' command on an API endpoint. Used typically to test API Gateway resources.
+
+By default, these assertions will check that the underlying call succeeds. That behavior can be extended
+to also include verification of the response.
+
+```ts
+test.assertAws(
+  s3.getObject({ ... }).returns({ Data: ... });
+);
+
+test.invokeLambda(...).returns({ ... });
+test.invokeLambda(...).throws({ ... });
+```
+
+When canned assertions are insufficient, `contest` also provides a mechanism to write custom assertions.
+Custom assertions are simply AWS Lambda Functions, and in the CDK, these are any construct that implements `IFunction`.
+
+```ts
+import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
+
+// Checks that a standard header is present in the given S3 object
+class StdHeaderAssertion extends Function {
+  constructor(scope: Construct, id: string, props: StdHeaderAssertionProps) {
+    super(scope, id, {
+      code: Code.fromAsset('./standard-header-assertion.js'),
+      runtime: Runtime.NODEJS_12_X,
+      env: {
+        BUCKET_NAME: props.bucket.bucketName,
+        KEY: props.bucket.key
+      },
+    });
+  }
+}
+```
+
+The assertion passes or fails based on the success or failure of the Lambda's handler code.
+In both cases, it can also return an optional message, that will be made available in the test result.
+
+To invoke this assertion, you can simply use the `invokeLambda()` API -
+
+```ts
+const hdrAssertion = new StdHeaderAssertion(...);
+test.invokeLambda(hdrAssertion);
+```
+
+### Test Execution & Report
+
+To execute your contest suite, you need to simply run `cdk contest` using the standard AWS CDK CLI.
+
+For projects that were created prior to the introduction of contest, or were not created using `cdk init`,
+a section for `contest` will need to be added to the `cdk.json` file of your project.
+The following applies to a CDK project in javascript that uses Node.js -
+
+```json
+{
+  "contest": {
+    "exec": "node {}",
+    "testRegex": "**/*.contest.js"
+  }
+}
+```
+
+- The `testRegex` is the glob pattern to find contest test files.
+- The `exec` key specifies the binary that should be invoked to synthesize a contest test.
+  The `{}` placeholder is required. This will be replaced by each file resolved by the `testRegex` glob.
+
+`cdk contest` will begin by discovering files and executing them one by one.
+As each test is completed, a one line summary of its pass/fail is printed to console.
+Detailed test results are available in a file, usually at `contest.out/results-xxx`.
+
+<!--
+
+#### TBD
+
+- In the pipeline
+- Snapshots
```

Comment:
This is intrinsic to the whole model: Looking for the CLI workflow around updating snapshots (basically something similar to `cdk-integ` and/or how the workflow is implemented in construct hub.